src/lj_vmevent_printer.c

```diff
@@ -0,0 +1,323 @@
+  }
+  int abortreason = tvisnumber(J->L->top-1) ? numberVint(J->L->top-1) : -1;
+  vmevent_log(context, "ABORT(%d): reason %d, %s:%d\n", T->traceno, abortreason, proto_chunknamestr(pt), lj_debug_line(pt, pc));
+}
+
+static void vmevent_traceflush(VMPrintContext *context, jit_State *J, uintptr_t data)
+{
+  vmevent_log(context, "TRACEFLUSH: mcode total %d\n", (uint32_t)J->szallmcarea);
+}
+
+static const char *getgcsname(int gcs)
+{
+  switch (gcs)
+  {
+  case GCSpause:
+    return "Pause";
+  case GCSpropagate:
+    return "Propagate";
+  case GCSatomic:
+    return "Atomic";
+  case GCSsweepstring:
+    return "SweepString";
+  case GCSsweep:
+    return "Sweep";
+  case GCSfinalize:
+    return "Finalize";
+  default:
+    return "?";
+  }
+}
+
+static void vmevent_gcstate(VMPrintContext *context, jit_State *J, int newstate)
+{
+  if (context->user.filter & EVENT_CLASS_GC) {
+    return;
+  }
+  vmevent_log(context, "GCSTATE: %s, totalmem %llukb, strings %d\n", getgcsname(newstate), (long long int)(J2G(J)->gc.total/1024), J2G(J)->strnum);
+}
+
+LUA_API void vmevent_printer_shutdown(lua_State *L);
+
+static void vmevent_protoloaded(VMPrintContext *context, jit_State *J, GCproto *pt)
+{
+  if (context->user.filter & EVENT_CLASS_PROTO_LOADED) {
+    return;
+  }
+  vmevent_log(context, "PROTO_LOADED(%p): %s:%d\n", pt, proto_chunknamestr(pt), pt->firstline);
+}
+
+static void free_context(VMPrintContext *context);
+
+static void vmevent_callback(void *contextptr, lua_State *L, int eventid, void *eventdata)
+{
+  VMEvent2 event = (VMEvent2)eventid;
+  VMPrintContext *context = contextptr;
+  jit_State *J = L2J(L);
+
+  switch (event) {
+    case VMEVENT_TRACE_START:
+      vmevent_tracestart(context, J, (GCtrace*)eventdata);
+      break;
+    case VMEVENT_RECORD:
+      vmevent_tracebc(context, J, eventdata);
+      break;
+    case VMEVENT_TRACE_STOP:
+      vmevent_tracestop(context, J, (GCtrace*)eventdata);
+      break;
+    case VMEVENT_TRACE_ABORT:
+      vmevent_traceabort(context, J, (GCtrace*)eventdata);
+      break;
+    case VMEVENT_TRACE_EXIT:
+      vmevent_exit(context, J, eventdata);
+      break;
+    case VMEVENT_BC:
+      vmevent_protoloaded(context, J, (GCproto*)eventdata);
+      break;
+    case VMEVENT_PROTO_BLACKLISTED:
+      vmevent_protobl(context, J, (VMEventData_ProtoBL*)eventdata);
+      break;
+    case VMEVENT_TRACE_FLUSH:
+      vmevent_traceflush(context, J, (int)(uintptr_t)eventdata);
+      break;
+    case VMEVENT_GC_STATECHANGE:
+      vmevent_gcstate(context, J, (int)(uintptr_t)eventdata);
+      break;
+    case VMEVENT_DETACH:
+      free_context(context);
+      break;
+    case VMEVENT_STATE_CLOSING:
+      if (J->vmevent_cb == vmevent_callback) {
+        luaJIT_vmevent_sethook(L, NULL, NULL);
+      }
+      free_context(context);
+      break;
+    default:
+      break;
+  }
+}
+
+static void default_ouput(VMPrintUserContext* ctx, const char *msg, uint32_t msglen)
+{
+  printf("%s", msg);
```

Comment:
could this be conditionally compiled?  i'd like to set something in the makefile or `luaconf.h` and avoid linking stdio in the JIT core.

Reply:
Luajit already uses stdio its the only way the lua based vmevent system can signal an uncaught Lua error happened in a vmevent hook.